File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/handlers/HealthRecordHandlerRegistry.kt
package com.phamtunglam.health_connector_hc_android.handlers

import androidx.annotation.VisibleForTesting
import androidx.health.connect.client.HealthConnectClient
import com.phamtunglam.health_connector_hc_android.DispatcherProvider
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.ActiveEnergyBurnedHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BasalBodyTemperatureHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BloodGlucoseHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BloodPressureHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BodyFatPercentageHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BodyTemperatureHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BodyWaterMassHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.BoneMassHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.CervicalMucusHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.CyclingPedalingCadenceHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.DistanceHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.ExerciseSessionHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.FloorsClimbedHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.HeartRateHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.HeartRateVariabilityRMSSDHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.HeightHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.HydrationHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.IntermenstrualBleedingHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.LeanBodyMassHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.MenstrualFlowInstantHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.MindfulnessSessionHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.NutritionHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.OvulationTestHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.OxygenSaturationHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.PowerSeriesHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.RespiratoryRateHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.RestingHeartRateHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.SexualActivityHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.SleepSessionHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.SpeedSeriesHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.StepsHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.TotalEnergyBurnedHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.Vo2MaxHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.WeightHandler
import com.phamtunglam.health_connector_hc_android.handlers.health_record_handlers.WheelchairPushesHandler
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto

/**
 * Central registry for Health Connect type handlers.
 */
internal class HealthRecordHandlerRegistry(
    private val dispatchers: DispatcherProvider,
    private val client: HealthConnectClient,
) {
    @get:VisibleForTesting
    internal val registeredHandlersCount: Int
        get() = handlers.size

    private val handlers: Map<HealthDataTypeDto, HealthRecordHandler> by lazy {
        buildMap {
            register(
                HeartRateHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                CyclingPedalingCadenceHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                HeightHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BodyTemperatureHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BasalBodyTemperatureHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                CervicalMucusHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                LeanBodyMassHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BodyFatPercentageHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                StepsHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                ActiveEnergyBurnedHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                TotalEnergyBurnedHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                DistanceHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                WeightHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                RestingHeartRateHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                OxygenSaturationHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                RespiratoryRateHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                Vo2MaxHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BloodGlucoseHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BloodPressureHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                FloorsClimbedHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                ExerciseSessionHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                HydrationHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                NutritionHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                OvulationTestHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                IntermenstrualBleedingHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                MenstrualFlowInstantHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                PowerSeriesHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                SexualActivityHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                SleepSessionHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                MindfulnessSessionHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                SpeedSeriesHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                WheelchairPushesHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BoneMassHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                BodyWaterMassHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
            register(
                HeartRateVariabilityRMSSDHandler(
                    dispatcher = dispatchers.io,
                    client = client,
                ),
            )
        }
    }

    /**
     * Retrieves a record handler for the specified health data type.
     *
     * @param type The health data type
     * @return The record handler for the type, or null if not found
     */
    fun getRecordHandler(type: HealthDataTypeDto): HealthRecordHandler? = handlers[type]

    /**
     * Registers a handler in the map builder.
     *
     * This is a private extension function used during map initialization.
     *
     * @param handler The handler to register
     */
    private fun MutableMap<HealthDataTypeDto, HealthRecordHandler>.register(
        handler: HealthRecordHandler,
    ) {
        this[handler.dataType] = handler
    }
}

File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/handlers/HealthRecordHandlerRegistry.swift
import Foundation
import HealthKit

/// Central registry mapping `HealthDataTypeDto` to type handlers
///
/// Uses `NSLock` to protect the handler dictionary for thread-safe concurrent initialization.
/// NSLock is used instead of Mutex for iOS 15+ compatibility (Mutex requires iOS 18+).
/// Marked @unchecked Sendable because NSLock-based synchronization ensures thread safety manually.
final class HealthRecordHandlerRegistry: @unchecked Sendable {
    /// Lock for thread-safe registry initialization
    private let initLock = NSLock()

    /// Storage for registered handler instances
    ///
    /// Key: `HealthDataTypeDto` (enum case)
    /// Value: Handler instance (any HealthRecordHandler)
    ///
    /// **Thread Safety:** Access protected by NSLock
    private var handlers: [HealthDataTypeDto: any HealthRecordHandler] = [:]

    /// The HealthKit store shared across all handlers
    private let healthStore: HKHealthStore

    /// Initializer
    ///
    /// - Parameter healthStore: The HKHealthStore to inject into all handlers
    init(healthStore: HKHealthStore) {
        self.healthStore = healthStore

        // Register all handlers (within initLock for thread safety)
        initLock.lock()
        defer { initLock.unlock() }

        registerAllHandlers()
    }

    /// Gets a handler with a specific capability (type-safe retrieval)
    ///
    /// - Parameters:
    ///   - type: The HealthDataTypeDto to look up
    ///   - capability: The protocol type the handler must conform to
    /// - Returns: The handler instance conforming to the capability
    /// - Throws: HealthConnectorError.unsupportedOperation if handler not found or lacks capability
    ///
    /// ## Example
    /// ```swift
    /// let handler = try registry.handler(for: .steps, withCapability: ReadableHealthRecordHandler.self)
    /// ```
    func handler<T>(
        for type: HealthDataTypeDto,
        withCapability capability: T.Type
    ) throws -> T {
        guard let baseHandler = handlers[type] else {
            throw HealthConnectorError.unsupportedOperation(
                message: "Unregistered handler for data type: \(type.rawValue)",
                context: ["dataType": type.rawValue]
            )
        }

        guard let handler = baseHandler as? T else {
            throw HealthConnectorError.unsupportedOperation(
                message:
                "Data type \(type.rawValue) does not support \(String(describing: capability))",
                context: [
                    "dataType": type.rawValue,
                    "capability": String(describing: capability),
                ]
            )
        }

        return handler
    }

    /// Register all type handlers
    ///
    /// This method is called once during initialization.
    /// Add new handlers here as they are implemented.
    ///
    /// **Note:** Must be called within initLock
    private func registerAllHandlers() {
        register(StepsHandler(healthStore: healthStore))
        register(ActiveEnergyBurnedHandler(healthStore: healthStore))
        register(BasalEnergyBurnedHandler(healthStore: healthStore))
        register(FloorsClimbedHandler(healthStore: healthStore))
        register(WheelchairPushesHandler(healthStore: healthStore))
        register(ElectrodermalActivityHandler(healthStore: healthStore))
        register(HydrationHandler(healthStore: healthStore))
        register(InsulinDeliveryHandler(healthStore: healthStore))
        register(WeightHandler(healthStore: healthStore))
        register(HeightHandler(healthStore: healthStore))
        register(BodyFatPercentageHandler(healthStore: healthStore))
        register(BodyTemperatureHandler(healthStore: healthStore))
        register(BasalBodyTemperatureHandler(healthStore: healthStore))
        register(SleepingWristTemperatureHandler(healthStore: healthStore))
        register(CervicalMucusHandler(healthStore: healthStore))
        register(LeanBodyMassHandler(healthStore: healthStore))
        register(HeartRateHandler(healthStore: healthStore))
        register(CyclingPedalingCadenceHandler(healthStore: healthStore))
        register(RestingHeartRateHandler(healthStore: healthStore))
        register(WalkingHeartRateAverageHandler(healthStore: healthStore))
        register(SexualActivityHandler(healthStore: healthStore))
        register(SleepStageHandler(healthStore: healthStore))
        register(MindfulnessSessionHandler(healthStore: healthStore))
        register(ExerciseSessionHandler(healthStore: healthStore))
        register(DietaryEnergyConsumedHandler(healthStore: healthStore))
        register(DietaryCaffeineHandler(healthStore: healthStore))
        register(DietaryProteinHandler(healthStore: healthStore))
        register(DietaryTotalCarbohydrateHandler(healthStore: healthStore))
        register(DietaryTotalFatHandler(healthStore: healthStore))
        register(DietarySaturatedFatHandler(healthStore: healthStore))
        register(DietaryMonounsaturatedFatHandler(healthStore: healthStore))
        register(DietaryPolyunsaturatedFatHandler(healthStore: healthStore))
        register(DietaryCholesterolHandler(healthStore: healthStore))
        register(DietaryFiberHandler(healthStore: healthStore))
        register(DietarySugarHandler(healthStore: healthStore))
        register(DietaryVitaminAHandler(healthStore: healthStore))
        register(DietaryVitaminB6Handler(healthStore: healthStore))
        register(DietaryVitaminB12Handler(healthStore: healthStore))
        register(DietaryVitaminCHandler(healthStore: healthStore))
        register(DietaryVitaminDHandler(healthStore: healthStore))
        register(DietaryVitaminEHandler(healthStore: healthStore))
        register(DietaryVitaminKHandler(healthStore: healthStore))
        register(DietaryThiaminHandler(healthStore: healthStore))
        register(DietaryRiboflavinHandler(healthStore: healthStore))
        register(DietaryNiacinHandler(healthStore: healthStore))
        register(DietaryFolateHandler(healthStore: healthStore))
        register(DietaryBiotinHandler(healthStore: healthStore))
        register(DietaryPantothenicAcidHandler(healthStore: healthStore))
        register(DietaryCalciumHandler(healthStore: healthStore))
        register(DietaryIronHandler(healthStore: healthStore))
        register(DietaryMagnesiumHandler(healthStore: healthStore))
        register(DietaryManganeseHandler(healthStore: healthStore))
        register(DietaryPhosphorusHandler(healthStore: healthStore))
        register(DietaryPotassiumHandler(healthStore: healthStore))
        register(DietarySeleniumHandler(healthStore: healthStore))
        register(DietarySodiumHandler(healthStore: healthStore))
        register(DietaryZincHandler(healthStore: healthStore))
        register(OxygenSaturationHandler(healthStore: healthStore))
        register(RespiratoryRateHandler(healthStore: healthStore))
        register(Vo2MaxHandler(healthStore: healthStore))
        register(BloodGlucoseHandler(healthStore: healthStore))
        register(NutritionHandler(healthStore: healthStore))
        register(OvulationTestHandler(healthStore: healthStore))
        register(IntermenstrualBleedingHandler(healthStore: healthStore))
        register(BodyMassIndexHandler(healthStore: healthStore))
        register(MenstrualFlowHandler(healthStore: healthStore))
        register(WaistCircumferenceHandler(healthStore: healthStore))
        register(HeartRateVariabilitySDNNHandler(healthStore: healthStore))
        register(BloodPressureHandler(healthStore: healthStore))
        register(SystolicBloodPressureHandler(healthStore: healthStore))
        register(DiastolicBloodPressureHandler(healthStore: healthStore))
        register(CyclingDistanceHandler(healthStore: healthStore))
        register(CyclingPowerHandler(healthStore: healthStore))
        register(RunningPowerHandler(healthStore: healthStore))
        register(SwimmingDistanceHandler(healthStore: healthStore))
        register(WheelchairDistanceHandler(healthStore: healthStore))
        register(WalkingRunningDistanceHandler(healthStore: healthStore))
        register(DownhillSnowSportsDistanceHandler(healthStore: healthStore))
        register(RowingDistanceHandler(healthStore: healthStore))
        register(PaddleSportsDistanceHandler(healthStore: healthStore))
        register(CrossCountrySkiingDistanceHandler(healthStore: healthStore))
        register(SkatingSportsDistanceHandler(healthStore: healthStore))
        register(SixMinuteWalkTestDistanceHandler(healthStore: healthStore))
        register(WalkingSpeedHandler(healthStore: healthStore))
        register(RunningSpeedHandler(healthStore: healthStore))
        register(StairAscentSpeedHandler(healthStore: healthStore))
        register(StairDescentSpeedHandler(healthStore: healthStore))
        register(PregnancyTestHandler(healthStore: healthStore))
        register(ProgesteroneTestHandler(healthStore: healthStore))
        register(LactationHandler(healthStore: healthStore))
        register(PregnancyHandler(healthStore: healthStore))
        register(ContraceptiveHandler(healthStore: healthStore))
        register(AlcoholicBeveragesHandler(healthStore: healthStore))
        register(ExerciseTimeHandler(healthStore: healthStore))
        register(MoveTimeHandler(healthStore: healthStore))
        register(StandTimeHandler(healthStore: healthStore))
        register(WalkingSteadinessHandler(healthStore: healthStore))
        register(WalkingAsymmetryPercentageHandler(healthStore: healthStore))
        register(WalkingDoubleSupportPercentageHandler(healthStore: healthStore))
        register(WalkingStepLengthHandler(healthStore: healthStore))
        register(SwimmingStrokesHandler(healthStore: healthStore))
        register(BloodAlcoholContentHandler(healthStore: healthStore))
        register(PeripheralPerfusionIndexHandler(healthStore: healthStore))
        register(ForcedVitalCapacityHandler(healthStore: healthStore))
        register(ForcedExpiratoryVolumeHandler(healthStore: healthStore))
        register(HighHeartRateEventRecordHandler(healthStore: healthStore))
        register(IrregularHeartRhythmEventRecordHandler(healthStore: healthStore))
        register(LowHeartRateEventRecordHandler(healthStore: healthStore))
        register(WalkingSteadinessEventRecordHandler(healthStore: healthStore))
        register(InfrequentMenstrualCycleEventRecordHandler(healthStore: healthStore))
        register(IrregularMenstrualCycleEventRecordHandler(healthStore: healthStore))
        register(PersistentIntermenstrualBleedingEventRecordHandler(healthStore: healthStore))
        register(ProlongedMenstrualPeriodEventRecordHandler(healthStore: healthStore))
        register(AtrialFibrillationBurdenHandler(healthStore: healthStore))
        register(NumberOfTimesFallenHandler(healthStore: healthStore))
    }

    /// Register a handler instance (called during init only)
    ///
    /// - Parameter handler: The handler instance to register
    ///
    /// **Note:** Must be called within initLock
    private func register(_ handler: any HealthRecordHandler) {
        handlers[type(of: handler).dataType] = handler
    }
}
